File: Python/get_aa874.py
# coding:utf-8

import urllib.request
import os
import re


basedir = 'D:\\aa874_1\\'
if not os.path.exists(basedir):
    os.makedirs(basedir)


def getHtmlContent(url):
    u = urllib.request.URLopener()
    u.addheaders = []
    u.addheader(
        'User-Agent', 'Opera/9.80 (Windows NT 6.1; WOW64; U; de) Presto/2.10.289 Version/12.01')
    u.addheader('Accept-Language', 'de-DE,de;q=0.9,en;q=0.8')
    u.addheader('Accept', 'text/html, application/xml;q=0.9, application/xhtml+xml, image/png, image/webp, image/jpeg, image/gif, image/x-xbitmap, */*;q=0.1')
    f = u.open(url)
    content = f.read().decode('UTF-8')
    print("网页解析成功")
    f.close()
    return content


def getUrl(html):
    num = 1
    # 解析第一层URL信息,通过尾页来判断总页数
    #<a href = "190.htm" > 尾页 < /a >
    first_flag = re.compile('<a href=\"(.*?)\">')
    first_url = first_flag.findall(html)
    # print(first_url[-1]) 取出总页数
    filepath, tempfilename = os.path.split(first_url[-1])
    shortname, ext = os.path.splitext(tempfilename)
    for i in range(1, int(shortname)):
        pre_url = ("https://aa874.com/htm/piclist9/%s.htm" % i)
        # print(pre_url) 取出每页的帖子信息
        html = getHtmlContent(pre_url)
        # 取出每帖图片信息
        pre_flag = re.compile('<li><a href=\"(.*?)\" target=\"_blank\">')
        pre_infos = pre_flag.findall(html)
        for m in pre_infos:
            # 构建图片真实下载地址
            sec_url = ("https://aa874.com%s" % m)
            sec_html = getHtmlContent(sec_url)
            sec_flag = re.compile('src=\"(.*?)\"><br>')
            sec_infos = sec_flag.findall(sec_html)
            for n in sec_infos:
                print("正在下载第%s张... %s" % (num, n))
                # 伪造user-agent通过采集限制
                img_opener = urllib.request.build_opener()
                img_opener.addheaders = [('User-agent', 'Mozilla/5.0')]
                fp = img_opener.open(n)
                img_data = fp.read()
                f = open(basedir + '%s.jpg' % num, 'w+b')
                f.write(img_data)
                f.close()
                fp.close()
                # urllib.request.urlretrieve(n, basedir + '%s.jpg' % num)
                num = num + 1
    print("一共下载了%s张图片！" % num)

if __name__ == "__main__":
    url = "https://aa874.com/htm/piclist1/"
    html = getHtmlContent(url)
    getUrl(html)

File: Go/get_aa874.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"

	"github.com/parnurzeal/gorequest"
)

const (
	PATH string = "/aa874"
)

var (
	c1 chan string
	c2 chan string
	c3 chan string
)

func init() {
	c1 = make(chan string, 1000) //page Url
	c2 = make(chan string, 1000) //tiezi Url
	c3 = make(chan string, 1000) //image url

}

func getHtmlContent(url string) (content string) {
	request := gorequest.New()
	_, body, _ := request.Get(url).
		Set("User-Agent", "Opera/9.80 (Windows NT 6.1; WOW64; U; de) Presto/2.10.289 Version/12.01").
		End()
	return string(body)
}

func getImage(url string) (image_body []byte) {
	request := gorequest.New()
	resq, _, _ := request.Get(url).
		Set("User-Agent", "Opera/9.80 (Windows NT 6.1; WOW64; U; de) Presto/2.10.289 Version/12.01").
		End()
	str, err := ioutil.ReadAll(resq.Body)
	checkErr(err)
	image_body = str
	return image_body

}

func get_page_info() {
	for i := 0; i <= 30; i++ {
		url_first := "https://aa874.com/htm/piclist9/" + strconv.Itoa(i) + ".htm"
		c1 <- url_first

	}
	close(c1)

}

func get_tiezi_url(c chan string) {
	reg, err := regexp.Compile(`<li><a href=\"(.*?)\" target=\"_blank\">`)
	checkErr(err)
	for page := range c {
		tiezi_content := getHtmlContent(page)
		tiezis := reg.FindAllStringSubmatch(tiezi_content, 1000)
		for _, tiezi_info := range tiezis {
			tiezi_url := "https://aa874.com" + tiezi_info[len(tiezi_info)-1]
			c2 <- tiezi_url
		}
	}
}

func get_tiezi_image(c chan string) {
	reg, err := regexp.Compile(`src=\"(.*?)\"><br>`)
	checkErr(err)
	for image_info := range c {
		images_content := getHtmlContent(image_info)
		images := reg.FindAllStringSubmatch(images_content, 1000)
		for _, image_info := range images {
			image_url := image_info[len(image_info)-1]
			c3 <- image_url
		}
	}
}

func Downloading(c chan string) {
	fnum := 1
	info, err := os.Stat(PATH)
	if err != nil || info.IsDir() == false {
		err := os.Mkdir(PATH, os.ModePerm)
		checkErr(err)
	}
	for image_url := range c {
		//		fmt.Println(image_url)
		image_body := getImage(image_url)
		filename := PATH + "/" + strconv.Itoa(fnum) + ".jpg"
		ioutil.WriteFile(filename, image_body, 0644)
		fmt.Println("Downloading..." + strconv.Itoa(fnum) + " Pics")
		fnum++
	}

}

func checkErr(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	go get_page_info()
	get_tiezi_url(c1)
	go get_tiezi_image(c2)
	Downloading(c3)

}
